### ISSUE TITLE
Let's update the tables to uchardet's

### ISSUE THREAD

**Issue by msdobrescu** [CharsetDetector/UTF-unknown#43]
Hello, let's update the tables to uchardet's ones as it seems to detect more cases.
I am willing to contribute the code, if you agree, except for the unit testing.

**Comment by 304NotModified** [comment]
That would be great!

PS, i need to fix this first, as the current unit tests (from the fork), don't test a thing....

https://github.com/CharsetDetector/UTF-unknown/pull/40

**Comment by msdobrescu** [comment]
There are things I need to clarify first. For instance, why some tables are exactly as uchardev's, but not all for the same language? What to do, upgrade them to uchrdev's anyway? Also naming seems to be an issue, like Latin1*something, when they could be ISO8859_1*something...? And charsets constants instead of hardcoded strings. Do you accept to have uchardev's notation's, like ILL, CTR etc.? That would be great. I can provide you a sample, on a branch, of what I did for my project, so far. A communication channel would be meeded too.

**Comment by 304NotModified** [comment]
> For instance, why some tables are exactly as uchardev's, but not all for the same language?

Do you have an example? This code is a fork of a fork, so have to investigate that. 



> So please wait on #40, I could give you a signal when it's done (or we could rebase the PR later on)

this is done

> like ILL, CTR etc

As longs as those are internal, it's fine to me

> I can provide you a sample, on a branch, of what I did for my project, so far.

👍 



> A communication channel would be meeded too.

Do you mean, chat?

**Comment by msdobrescu** [comment]
Just a sample. Given #define ILL 255, CTR 254, SYM 253, RET 252, NUM 251 in uchardet, for Hebrew, the model table is the same. But, for windows-1255, the character mapping table differs:
UTF-unknown:

`private readonly static byte[] WIN1255_CHAR_TO_ORDER_MAP = {
            255,255,255,255,255,255,255,255,255,255,254,255,255,254,255,255,  //00
            255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,  //10
            +253,253,253,253,253,253,253,253,253,253,253,253,253,253,253,253,  //20
            252,252,252,252,252,252,252,252,252,252,253,253,253,253,253,253,  //30
            253, 69, 91, 79, 80, 92, 89, 97, 90, 68,111,112, 82, 73, 95, 85,  //40
            78,121, 86, 71, 67,102,107, 84,114,103,115,253,253,253,253,253,  //50
            253, 50, 74, 60, 61, 42, 76, 70, 64, 53,105, 93, 56, 65, 54, 49,  //60
            66,110, 51, 43, 44, 63, 81, 77, 98, 75,108,253,253,253,253,253,  //70
            124,202,203,204,205, 40, 58,206,207,208,209,210,211,212,213,214,
            215, 83, 52, 47, 46, 72, 32, 94,216,113,217,109,218,219,220,221,
            34,116,222,118,100,223,224,117,119,104,125,225,226, 87, 99,227,
            106,122,123,228, 55,229,230,101,231,232,120,233, 48, 39, 57,234,
            30, 59, 41, 88, 33, 37, 36, 31, 29, 35,235, 62, 28,236,126,237,
            238, 38, 45,239,240,241,242,243,127,244,245,246,247,248,249,250,
            9,  8, 20, 16,  3,  2, 24, 14, 22,  1, 25, 15,  4, 11,  6, 23,
            12, 19, 13, 26, 18, 27, 21, 17,  7, 10,  5,251,252,128, 96,253,
        };`

uchardet:

`static const unsigned char win1255_CharToOrderMap[] =
{
CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,RET,CTR,CTR,RET,CTR,CTR,  //00
CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,CTR,  //10
SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,SYM,  //20
NUM,NUM,NUM,NUM,NUM,NUM,NUM,NUM,NUM,NUM,SYM,SYM,SYM,SYM,SYM,SYM,  //30
SYM, 69, 91, 79, 80, 92, 89, 97, 90, 68,111,112, 82, 73, 95, 85,  //40
 78,121, 86, 71, 67,102,107, 84,114,103,115,SYM,SYM,SYM,SYM,SYM,  //50
SYM, 50, 74, 60, 61, 42, 76, 70, 64, 53,105, 93, 56, 65, 54, 49,  //60
 66,110, 51, 43, 44, 63, 81, 77, 98, 75,108,SYM,SYM,SYM,SYM,SYM,  //70
124,ILL,203,204,205, 40, 58,206,207,208,ILL,210,ILL,ILL,ILL,ILL,
ILL, 83, 52, 47, 46, 72, 32, 94,216,113,ILL,109,ILL,ILL,ILL,ILL,
 34,116,222,118,100,223,224,117,119,104,125,225,226, 87, 99,227,
106,122,123,228, 55,229,230,101,231,232,120,233, 48, 39, 57,234,
 30, 59, 41, 88, 33, 37, 36, 31, 29, 35,235, 62, 28,236,126,237,
238, 38, 45,239,240,241,242,243,127,ILL,ILL,ILL,ILL,ILL,ILL,ILL,
  9,  8, 20, 16,  3,  2, 24, 14, 22,  1, 25, 15,  4, 11,  6, 23,
 12, 19, 13, 26, 18, 27, 21, 17,  7, 10,  5,ILL,ILL,128, 96,ILL,
};`

Regarding the chat, what are the options? Maybe e-mail too?

Anyway, I have no rights to create even a branch...

**Comment by msdobrescu** [comment]
One more thing. I don't trust much the framework regarding optimizations.
From my experience, it is a good practice to declare some variable outside a loop, initialize it with a default, then use it inside instead of re-declaring it inside the loop. This avoids memory fragmentation.

**Comment by msdobrescu** [comment]
Hi, I still can't make a branch.

**Comment by 304NotModified** [comment]
Hi, you need to create a fork, create a branch, make changes, push and than create a PR :)

**Comment by 304NotModified** [comment]
see also https://help.github.com/articles/fork-a-repo/

**Comment by msdobrescu** [comment]
And if it's OK for, would you integrate it back?

**Comment by 304NotModified** [comment]
yes of course :)

**Comment by msdobrescu** [comment]
Hi, a fork is created, I have added a branch with the first bunch of changes, related to the sample size that is no more a const of 64, but a member of a class, in order to accept different sample sizes, as they come from the uchardet library.

**Comment by msdobrescu** [comment]
Hello, any chance to look at it?
This is the first commit, I'd start with three, so more to come.
I see a Greek test failing in github, but locally, on my computer, passes.

**Comment by rstm-sf** [comment]
Hello all!

@msdobrescu, why did you choose the implementation by [_uchardet_](https://www.freedesktop.org/wiki/Software/uchardet/)?

I ask because I recently compared this implementation with the _uchardet_ on specific data with little text. And the _uchardet_ was worse on russian encodings.

**Comment by msdobrescu** [comment]
I didn't know that. uchardet is more complete and flexible.
Its statistics data component must be the reason of your problem.
Need to have some samples in order to see what happens.
The more text you have in a file, the safest is the detection, because having some encoding specific characters occurring in a language pattern fashion reveals the actual encoding better.
Here is just a port attempt of an actively developed library on this matter. Did you try to discuss and add bug reports of what you have encountered directly to the uchardet project boards?

**Comment by rstm-sf** [comment]
Thanks for the answer!

> uchardet is more complete and flexible.
Its statistics data component must be the reason of your problem.

Perhaps that's why. Since the data is really specific and may even contain abbreviations of words.

> Need to have some samples in order to see what happens.

Sorry, but I have no samples now.

> Did you try to discuss and add bug reports of what you have encountered directly to the uchardet project boards?

Unfortunately, I did not discuss.

I tried to compare the tables now and got as a result that only the selected and reassigned _uchardet_ codes (ILL, CTR, SYM, RET, NUM) are different. You can see [Python3 source code](https://gist.github.com/rstm-sf/7d19f1bbb1afebabd417aee4dcc59d88) (can use any online python compiler).

**Comment by rstm-sf** [comment]
I understand that only _ILL_ and _CTR_ are  important (see: [nsSingleByteCharSetProber::HandleData](https://gitlab.freedesktop.org/uchardet/uchardet/blob/master/src/nsSBCharSetProber.cpp#L45)).

I think that each model should have its own values ​​for the _Character Mapping Table_ (ILL, CTR, SYM, RET, NUM).

**Comment by rstm-sf** [comment]
> I think that each model should have its own values ​​for the _Character Mapping Table_ (ILL, CTR, SYM, RET, NUM).

This is a misconception. Now I began to understand why the need for variable _charToOrderMap_.

**Comment by rstm-sf** [comment]
Most likely because of [this](https://gitlab.freedesktop.org/uchardet/uchardet/blob/master/src/nsSBCharSetProber.cpp#L128), other language encodings rise in ranking.

**Comment by rstm-sf** [comment]
I found an [example](https://github.com/CharsetDetector/UTF-unknown/files/2905591/sample.txt) where UTF Unknown correctly identified encoding (cp855), and uchardet - no (utf-8).

**Comment by msdobrescu** [comment]
I don't get the same result as you, @rstm-sf . on my branch it identifies it as expected.

**Comment by msdobrescu** [comment]
As a general approach, Notepad++ uses uchardet, so you can expect the same results.
Also Notepad++ identifies the file as ibm855.

**Comment by rstm-sf** [comment]
> As a general approach, Notepad++ uses uchardet

Thanks for the interesting information.

> on my branch it identifies it as expected.

Yes, it's. I looked at the return value _Confidence_ and got that they are different:
- 0.8776797 (master branch)
- 0.685687244 (your branch)
---
I debuged my _uchardet_ usage:
- in this [for-loop](https://gitlab.freedesktop.org/uchardet/uchardet/blob/master/src/nsUniversalDetector.cpp#L317) return value _mCharSetProbers[i]->GetConfidence()_ is 0.752499998. And [then](https://gitlab.freedesktop.org/uchardet/uchardet/blob/master/src/nsUniversalDetector.cpp#L331),  value _m_charset_ is assigned [this](https://gitlab.freedesktop.org/uchardet/uchardet/blob/master/src/uchardet.cpp#L65) return [value](https://gitlab.freedesktop.org/uchardet/uchardet/blob/master/src/nsUTF8Prober.h#L51)

In your branch:
- for multi-byte prober [this](https://github.com/msdobrescu/UTF-unknown/blob/3b9135fc16cc3e876ec63bf6b2a4808e9226c1e6/src/Core/MBCSGroupProber.cs#L176) check is achieved only when GB18030Prober object with _cf = 0.01_
- single-byte prober: cp855 with _cf = 0.685687244

In master branch:
- multi-byte prober: GB18030Prober with _cf = 0.01_
- single-byte prober: cp855 with _cf = 0.8776797_

**Comment by msdobrescu** [comment]
The master branch [has a different calculation](https://github.com/msdobrescu/UTF-unknown/commit/504f723abae8ae90bb3cd074f11293845608c01f#diff-c6c9169620eaafcfebab7f031b7ae468R164) and different data for some of the various languages, so the confidence value is different for all.

**Comment by rstm-sf** [comment]
@msdobrescu, I don't mind different confidences :)
It was explained (first of all for myself) that the values ​​are, in fact, calculated the same, that for _UTF Unknown_, that for _uchardet_. Difference occurs only when determining the value confidence for _MBCSGroupProber_ object.

**Comment by msdobrescu** [comment]
I did not yet take the MBCS prober.

**Comment by rstm-sf** [comment]
@msdobrescu, are you going to take?

**Comment by msdobrescu** [comment]
For now, I add the single byte languages.
It's hallucinating, I need a way to check I did not forget some table...
Then I move to others. It's good to know I am not alone :)

**Comment by rstm-sf** [comment]
I tried to create an [issue](https://gitlab.freedesktop.org/uchardet/uchardet/issues/9) in the uchardet repository.

**Comment by msdobrescu** [comment]
> I tried to create an [issue](https://gitlab.freedesktop.org/uchardet/uchardet/issues/9) in the uchardet repository.
What makes you think one or another is correct or wrong?

**Comment by rstm-sf** [comment]
> What makes you think one or another is correct or wrong?

One example that showed that the result is incorrect.

The difference is in determining the level of confidence for the multi-byte encoding. Earlier in the _uchardet_ there was the same check, in [case](https://github.com/CharsetDetector/UTF-unknown/blob/d17ab7282a419de7313ffb90be3debfea32653c2/src/Library/Ude.Core/MBCSGroupProber.cs#L126) of such an [event](https://github.com/CharsetDetector/UTF-unknown/blob/d17ab7282a419de7313ffb90be3debfea32653c2/src/Library/Ude.Core/UTF8Prober.cs#L78). But, why the _uchardet_ refused this. Which ultimately gives the detection of utf-8. If I can understand why this is better, then I will accept that my example is an exception.

**Comment by 304NotModified** [comment]
I'm curious for the results. If the detection get worse, I prefer not to change it.

**Comment by msdobrescu** [comment]
Question is how do you decide it is worse? Being a lower value is not a criterium.
For instance, I have moved test files from windows-1252 to iso-8859-1 because they did not contain windows-1252 characters at all. When I have added such characters, they were detected as windows-1252. Same for the original uchardet. It does not matter if you create a windows-1252 file without the windows-1252 specific content. Let's see the arguments now. Why do you say one or another is wrong or right in the case of multibyte?

**Comment by rstm-sf** [comment]
> Question is how do you decide it is worse?

Maybe because I am a native speaker of Russian? And I want to believe in what I see? :)

According to the [table](https://en.wikipedia.org/wiki/Code_page_855), [all characters](https://github.com/CharsetDetector/UTF-unknown/files/2905591/sample.txt)  (```C7 A0 D0 A0 D2 C6 A0 E3```)  are decoded into letters.

**Comment by msdobrescu** [comment]
No, I mean, what is the problem with multibyte confidence value? Please explain why UTF-unknown is better than uchardet. Technically. Or what is the issue? I believe you have spotted something, just I need to go into details in order to assess the situation.

**Comment by rstm-sf** [comment]
Well, let's try to explain in another way.

I will consider on the example of the _Utf Unknown_.

In the case when the buffer (```C7 A0 D0 A0 D2 C6 A0 E3```) is entered, the variant _InputState.Highbyte_ is [considered](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/CharsetDetector.cs#L303).

Next, go to the [_multi-byte Prober_](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/MBCSGroupProber.cs#L56) (aka _charsetProbers[0]_, _charsetProbers_ init [here](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/CharsetDetector.cs#L349)). We are now interested in [how we calculate the confidence level](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/MBCSGroupProber.cs#L136), or more specifically: either [no multibyte encoding](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/MBCSGroupProber.cs#L142) or [find the best](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/MBCSGroupProber.cs#L144).

Next, go to the [_utf-8 Prober_](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/UTF8Prober.cs#L47). When we researching the buffer, then for each character we have a chance to get into an [error](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/UTF8Prober.cs#L73). And here at _i == 5_ we get that _codingState_ is equal to _SMModel.ERROR_. Which [allows us not to consider _utf-8 Prober_ (```isActive[i] = false;```)](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/MBCSGroupProber.cs#L124) when finding the confidence level.

And this check is missing in the _uchardet_. Therefore, we will know the confidence level for _utf-8 Prober_. And I don't know why this check was removed and what does [this line](https://github.com/CharsetDetector/UTF-unknown/blob/68d4d684cff73227c1b85f4994766675ac275e1d/src/Core/CodingStateMachine.cs#L69) mean. Perhaps they found a case that this check is not correct or showed that it is fundamentally wrong?

**Comment by msdobrescu** [comment]
Too bad I can't find Mozilla uchardet original source. Are there only forks? It seems the code was simply updated from the upstream and the reasons for this change are lost.

**Comment by rstm-sf** [comment]
In the [subject neighborhood](https://github.com/CharsetDetector/UTF-unknown/issues/30#issue-220332488), there is a link on the repository :)

There is also a [link](http://hg.mozilla.org/mozilla-central/file/tip/extensions/universalchardet/) to partially updated source code (I learned about it from [here](https://github.com/albfernandez/juniversalchardet)).

**Comment by msdobrescu** [comment]
Why the links were wrong before (lxr instead of dxr...).
Anyway, here is the change: https://hg.mozilla.org/mozilla-central/rev/976bb1d02042
It says "Remove unnecessary and obsolete code", apparently related to a bug: https://bugzilla.mozilla.org/show_bug.cgi?id=424916
I think this is the relevant part: "_Turning off the individual charset detectors when we encounter illegal sequences isn't really necessary. Cutting it out doesn't seem to cause either misdetections or a performance hit._"

**Comment by rstm-sf** [comment]
Wow thanks!

> Cutting it out doesn't seem to cause either misdetections or a performance hit.

Or does it seem? :)

But anyway, now I understand what it is connected with.

**Comment by msdobrescu** [comment]
In MBCS case, the uchardet port shows the following dump for the sample file:

```
Get confidence:
-- new match found: confidence 0.01, index 1, charset Shift-JIS.
Get confidence done.
 MBCS Group Prober --------begin status
 MBCS inactive: UTF-8 (i.e. confidence is too low).
 MBCS 0.01: [Shift-JIS]

 MBCS inactive: EUC-JP (i.e. confidence is too low).
 MBCS 0.01: [GB18030]

 MBCS inactive: EUC-KR (i.e. confidence is too low).
 MBCS inactive: Big5 (i.e. confidence is too low).
 MBCS inactive: EUC-TW (i.e. confidence is too low).
 MBCS Group found best match [Shift-JIS] confidence 0.01.
```

@rstm-sf, I see `MBCS inactive: UTF-8 (i.e. confidence is too low)`. What is wrong with that?

**Comment by rstm-sf** [comment]
> What is wrong with that?

It is good here, because under the condition of ```ProbingState.NotMe``` we [get](https://github.com/msdobrescu/UTF-unknown/blob/18c8a265c7e46136b53772bd9f1dce6c304e960d/src/Core/MBCSGroupProber.cs#L167) _0.01_ value.

Or otherwise, we [get](https://github.com/msdobrescu/UTF-unknown/blob/18c8a265c7e46136b53772bd9f1dce6c304e960d/src/Core/MBCSGroupProber.cs#L218) the message:
> MBCS inactive: UTF-8 (i.e. confidence is too low).
---
@msdobrescu, maybe I misunderstood you?

Besides the fact that you expand the detection of single-byte encodings I have no problems. It is very good that you do it.

In the beginning, I thought that the _uchardet_ library had problems with defining single-byte encoding. But in the end I came to the conclusion that the modified detection of multi-byte encoding proceeds. And now I'm glad you helped to understand it out.

**Comment by msdobrescu** [comment]
Do you think it's alright so far, gentlemen?

**Comment by rstm-sf** [comment]
Maybe refactoring file paths like [here](https://github.com/rstm-sf/UTF-unknown/tree/feature/prober_language_model/src/Core)?

**Comment by msdobrescu** [comment]
> Maybe refactoring file paths like [here](https://github.com/rstm-sf/UTF-unknown/tree/feature/prober_language_model/src/Core)?
I don't know, I don't like that unless I alter the namespaces accordingly.

**Comment by msdobrescu** [comment]
@304NotModified , what do you think of this change?
I will not reorganize the project files and namespaces until it is stable, because will make the review hard to follow.

**Comment by 304NotModified** [comment]
Thanks, we try to review this weekend. 

@rstm-sf do you think this is an improvement?

FYI I forked and refactored this code base so it's easier to use and maintain, but I think you guys have more knowledge of the domain :)

**Comment by msdobrescu** [comment]
No, I don't. I speak for myself only.
uchardet is simply more diverse and used widely.
So far I did not have issues, but anytime might raise some.

**Comment by rstm-sf** [comment]
> do you think this is an improvement?

Yes of course.

**Comment by msdobrescu** [comment]
Hi, do you need help or have you found something?

**Comment by 304NotModified** [comment]
Yes, please update the readme to mention uchardet, if not done yet. :)

This will be 2.0. 

Sorry for the delay, large PRs are time consuming

**Comment by 304NotModified** [comment]
merged :) Thanks for the great work @msdobrescu !

**Comment by msdobrescu** [comment]
Thanks _you_!
